source/opt/local_ssa_elim_pass.cpp

```diff
@@ -0,0 +1,759 @@
+    // add true successors
+    blk.ForEachSuccessorLabel([&blk, this](uint32_t sbid) {
+      block2structured_succs_[&blk].push_back(id2block_[sbid]);
+    });
+  }
+}
+
+void LocalSSAElimPass::ComputeStructuredOrder(
+    ir::Function* func, std::list<ir::BasicBlock*>* order) {
+  // Compute structured successors and do DFS
+  ComputeStructuredSuccessors(func);
+  auto ignore_block = [](cbb_ptr) {};
+  auto ignore_edge = [](cbb_ptr, cbb_ptr) {};
+  auto get_structured_successors = [this](const ir::BasicBlock* block) {
+      return &(block2structured_succs_[block]); };
+  // TODO(greg-lunarg): Get rid of const_cast by making moving const
+  // out of the cfa.h prototypes and into the invoking code.
+  auto post_order = [&](cbb_ptr b) {
+      order->push_front(const_cast<ir::BasicBlock*>(b)); };
+  
+  spvtools::CFA<ir::BasicBlock>::DepthFirstTraversal(
+      &pseudo_entry_block_, get_structured_successors, ignore_block,
+      post_order, ignore_edge);
+}
+
+void LocalSSAElimPass::SSABlockInitSinglePred(ir::BasicBlock* block_ptr) {
+  // Copy map entry from single predecessor
+  const uint32_t label = block_ptr->id();
+  const uint32_t predLabel = label2preds_[label].front();
+  assert(visitedBlocks_.find(predLabel) != visitedBlocks_.end());
+  label2ssa_map_[label] = label2ssa_map_[predLabel];
+}
+
+bool LocalSSAElimPass::IsLiveAfter(uint32_t var_id, uint32_t label) const {
+  // For now, return very conservative result: true. This will result in
+  // correct, but possibly usused, phi code to be generated. A subsequent
+  // DCE pass should eliminate this code.
+  // TODO(greg-lunarg): Return more accurate information
+  (void) var_id;
+  (void) label;
+  return true;
+}
+
+uint32_t LocalSSAElimPass::Type2Undef(uint32_t type_id) {
+  const auto uitr = type2undefs_.find(type_id);
+  if (uitr != type2undefs_.end())
+    return uitr->second;
+  const uint32_t undefId = TakeNextId();
+  std::unique_ptr<ir::Instruction> undef_inst(
+    new ir::Instruction(SpvOpUndef, type_id, undefId, {}));
+  def_use_mgr_->AnalyzeInstDefUse(&*undef_inst);
+  module_->AddGlobalValue(std::move(undef_inst));
+  type2undefs_[type_id] = undefId;
+  return undefId;
+}
+
+uint32_t LocalSSAElimPass::GetPointeeTypeId(
+    const ir::Instruction* ptrInst) const {
+  const uint32_t ptrTypeId = ptrInst->type_id();
+  const ir::Instruction* ptrTypeInst = def_use_mgr_->GetDef(ptrTypeId);
+  return ptrTypeInst->GetSingleWordInOperand(kTypePointerTypeIdInIdx);
+}
+
+void LocalSSAElimPass::SSABlockInitLoopHeader(
+    std::list<ir::BasicBlock*>::iterator block_itr) {
+  const uint32_t label = (*block_itr)->id();
+  // Determine backedge label.
+  uint32_t backLabel = 0;
+  for (uint32_t predLabel : label2preds_[label])
+    if (visitedBlocks_.find(predLabel) == visitedBlocks_.end()) {
+      assert(backLabel == 0);
+      backLabel = predLabel;
+      break;
+    }
+  assert(backLabel != 0);
+  // Determine merge block.
+  auto mergeInst = (*block_itr)->begin();
```

Comment:
Good catch.

This is only called on a loop header so it is safe to just reference the second-to-last instruction.